docs/src/appendix/packaging.md

```diff
@@ -0,0 +1,501 @@
+allowing efficient cross-language interop. We follow the semantics of the Wasm canonical ABI, but layered on top of
+our canonical Miden ABI, which describes how the Wasm abstract machine is translated to the Miden abstract machine.
+
+2. To provide the compiler with the necessary type information needed to emit lifting/lowering code at call boundaries,
+as well as mapping imports (from the packages' dependencies) to their corresponding MAST roots, so that calls use those
+roots rather than the fully-qualified names. This ensures that the code that was compiled and published is always the
+exact code that gets executed, in perpetuity.
+
+Because the type descriptor is in binary form, to view it in the more human-friendly WIT format, you
+will need to use [wasm-tools](https://github.com/bytecodealliance/wasm-tools), like so:
+
+    wasm-tools component wit extracted-package/component.wasm
+
+The contents of `component.wasm` here is not an actual Wasm module, but simply the binary representation
+of the WIT package.
+
+## Directory Layout
+
+The canonical layout of a package when extracted as a directory, is as follows:
+
+    name@version/
+    |- manifest.toml
+    |- component.wasm
+    |- lib/
+       |- 0x00000000.mast
+       |- 0x00000001.mast
+    |- rodata/
+       |- 0x0.bin
+    |- etc/
+       |- component.sourcemap
+       |- docs/
+          |- index.html
+       |- vendor/
+          |- foo@0.1.0.mpkg
+
+### Manifest
+
+The `manifest.toml` file contains all of the top-level metadata about the package. When produced as a result
+of extracting a package, it's contents may differ from the `manifest.toml` used to create the package. This
+is because we provide some helpful functionality in the manifest for creating packages by hand, that are
+translated into canonical form when the package is created. These "canonicalizations" are covered below
+in each relevant section.
+
+> [!NOTE]
+> TOML was chosen as the format for the manifest due to it being a nicer format for
+> writing by hand, while being strongly typed, unambiguous, and supported by good
+> tooling in most languages.
+
+The manifest begins with basic metadata about the package:
+
+```toml
+[package]
+name = "my-package"
+version = "0.1.0"
+miden-version = "0.8.0"  # Optional
+```
+
+The `miden-version` field lets you specify a minimum Miden VM version for the package. Any attempt to load
+a package with a lower-versioned VM will fail. This field is optional, and when not present, is equivalent
+to no minimum required version.
+
+#### Package Type
+
+There are two primary types of packages: libraries and programs (executables). These correspond to the
+`[lib]` and `[bin]` sections of the manifest, respectively. Only one section is permitted.
+
+For libraries:
+
+```toml
+[lib]
+start = "0x..." # Optional
```

Comment:
> Assuming that lib directory contains a MastForest, would we not also need to define which of the MAST roots define the public interface for the library?

I alluded to this in my reply to your comment about the WIT/component definition - but this is exactly what the component definition tells us. That's not to say that we can't provide this information another way, but we essentially kill two birds with one stone.

> Also, would I wonder if it would make sense to commit to the whole MAST forest somehow - similar to how entrypoint for executable programs commits to everything that can be executed by the program.

I don't think committing to the forest is necessary, since the MAST roots already provide that at a more granular level (and you likely only care about the parts of the package you actually use in practice). However, I do think there is benefit in committing to the package contents (or signing them and validating the signature), in order to ensure that when you fetch a package from some remote server (as an example), that you are getting the exact same package you requested, and not one that has been modified in some subtle way. Maybe that's what you were getting at though.

> Lastly, is start more for the future? Right now, Miden VM would not be able to process a package with the start program. Unless, we make an assumption that before any public procedure start program needs to be executed. But I'm not sure if it is a good assumption to make in the VM.

There are a few things here:

1. The `start` function is exactly that - a procedure which is executed before any other code in the same component, and is responsible for one-time initialization for a given instance of the component (i.e. it is run in each new instance of the component). This function does not have access to any other code in the component while it is being executed. The compiler will enforce this in packages it produces, but for now the VM doesn't need to care about this.
2. This is primarily useful for the compiler at this point in time, when compiling an executable, so that it can inject calls to the `start` function of any packages it depends on in the compiler-generated setup code in the entrypoint. It also, for the time being, will do this in the prologue of procedures which are `call`-able (same as with rodata initialization). Ideally though, this will be an aspect of component initialization that will be handled by the VM when it creates a new instance of a component, but obviously we don't have that functionality yet.
3. In the near term, I think a good approach to handling this would be for us to use an undocumented feature flag in executable packages to indicate that an executable was produced by our compiler, and then raise an error when any package with the `start` field set is loaded/used while executing a program _without_ that feature. That feels a bit hacky, but allows us to tell users when they accidentally use a package in a way that violates its expectations. Ultimately, we would be able to just support these packages when we no longer need to rely on the compiler to handle this, so this would be trivially backwards-compatible.

For now, `start` is primarily required to support rodata initialization in libraries compiled from Rust (as well as any Rust code that causes the `start` function to be emitted in Wasm). When we are not producing an executable, we have nowhere to place this kind of code, and no way other than this to signal to other tools that such initialization is needed/required, and how to do so. The `start` field gives us this, and on the VM side, if we choose to do so, provides us a way to recognize when a given package cannot be supported/executed correctly (with something akin to what I described in point 3 above in place).